Allow extensions to make buttons

cwillisf opened the pull request:
This PR depends on LLK/scratch-gui#4712 for the `MAKE_A_*` callback registrations.

### Proposed Changes

This change adds support for a "button" (pseudo-) block type to the extension interface. These entries are expected to have a `func` field but no `opcode`. For now the `func` must be one of a few existing callback keys handled by `scratch-blocks` but in the future we could open this up to allow connecting an extension method to a button.

### Reason for Changes

Converting variables, lists, and "My Blocks" into extensions means supporting the buttons used to create those items. This approach also allows us to experiment with other extension buttons in the future, if we'd like.

### Test Coverage

A unit test has been added to ensure that `BlockType.BUTTON` gets interpreted correctly.

Commit:
Allow extensions to make buttons

Commit:
Add unit test for extension button

towerofnix commented:
>Converting ... "My Blocks" into extensions

:eyes: This shows that extensions definitely have potential in this direction..!

kchadha commented on test/unit/extension_conversion.js:
```diff
@@ -12,6 +12,11 @@ const testExtensionInfo = {
     name: 'fake test extension',
     blocks: [
+        {
+            func: 'CREATE_VARIABLE',
```
could you also add this to the `coreExample` extension for easy testing? =D

Commit:
Fix missing arg for extension block separators

Commit:
Use new Scratch-specific callback keys for extension buttons

Commit:
Add a button to the CoreEx extension

kchadha commented on src/extension-support/extension-manager.js:
```diff
@@ -375,29 +375,36 @@ class ExtensionManager {
+            }
+            break;
+        default:
             blockInfo.func = blockInfo.func ? this._sanitizeID(blockInfo.func) : blockInfo.opcode;
```
Should we log a warning or an error if opcode is undefined here? Or if both opcode and func are undefined? Alternatively, if they are undefined, should we default to a built in no-op function? Maybe we do that somewhere else...

kchadha replied:
Ah, never mind, I see we throw an error below. Nevertheless, I'll leave the question about defaulting to a no-op function here.

cwillisf replied:
Good idea -- done. Note that the exception thrown here will be caught and reported by some code in `_prepareExtensionInfo` which dumps the whole `blockInfo` so there's no need to add any `blockInfo` fields in the exception message.

kchadha commented on test/integration/internal-extension.js:
```diff
@@ -83,21 +83,24 @@ test('load sync', t => {
 
     t.equal(vm.runtime._blockInfo.length, 1);
-    t.equal(vm.runtime._blockInfo[0].blocks.length, 1);
+    t.equal(vm.runtime._blockInfo[0].blocks.length, 2);
```
Maybe add a comment here that there are 2 items but only one actual block. The other one is a button.

kchadha reviewed (approved):
LGTM! Just some minor questions/comments.

Commit:
Explicitly check that every extension block has an opcode

Commit:
Test the `info` field in converted extension data

There are parts of the extension registration process which rely on the
`info` field being non-null, even for block separators. At one point
during development I broke that, so here's a test to hopefully prevent
someone else from getting bitten by the same thing.

Commit:
Add a comment clarifying the items in `_blockInfo[0].blocks`

kchadha commented on test/unit/extension_conversion.js:
```diff
@@ -153,9 +163,15 @@ test('registerExtensionPrimitives', t => {
 
+        blocksInfo.forEach(blockInfo => {
+            // `true` here means "either an object or a non-empty string but definitely not null or undefined"
+            t.true(blockInfo.info, 'Every block and pseudo-block must have a non-empty "info" field');
```
I think in other VM unit tests we use `t.ok` for the truthy test. I believe that's just a synonym for t.true... ¯\_(ツ)_/¯

cwillisf replied:
Yeah -- I felt like `t.true` was semantically more accurate in this case since usually we do `t.ok(something == desired)`, and in this case I just want to check that it's "truthy". Really I kind of want `t.typeIsOneOf(blockInfo.info, ['object','string']);` but that doesn't exist.

I guess it could be `t.ok(-1 != ['object', 'string'].indexOf(typeof blockInfo.info))` but that seems even harder to read than the other options :(

Also, obligatory "you dropped this: \\"

towerofnix replied:
`t.ok(['object', 'string'].includes(typeof blockInfo.info))` perhaps?

kchadha replied:
> Really I kind of want `t.typeIsOneOf(blockInfo.info, ['object','string']);` but that doesn't exist.

Agreed.

> I guess it could be `t.ok(-1 != ['object', 'string'].indexOf(typeof blockInfo.info))` but that seems even harder to read than the other options :(

Also agreed.

> Also, obligatory "you dropped this: \"

I think GitHub ate it! It's there if I try to edit the raw comment text but idk what markdown feature GitHub thinks I'm trying to use.
